File: backend/scripts/grant_counselor_credits_by_email.py
#!/usr/bin/env python3
"""
상담사 이메일로 검사 크레딧 지급.

사용 예 (저장소 루트, GOOGLE_APPLICATION_CREDENTIALS 설정 후):
  python backend/scripts/grant_counselor_credits_by_email.py jomiga@naver.com 100000
  python backend/scripts/grant_counselor_credits_by_email.py jomiga@naver.com jomigata@naver.com --amount 100000
"""
from __future__ import annotations

import argparse
import sys
from pathlib import Path

sys.path.insert(0, str(Path(__file__).resolve().parents[1]))

from firebase_admin import auth as fb_auth

from config import USERS_COLLECTION
from firebase_init import get_firestore, get_firebase_app
from utils.counselor_credits import get_balance, grant_credits

get_firebase_app()


def resolve_counselor_uid(db, email: str) -> tuple[str, str]:
    email = email.strip().lower()
    if not email:
        raise ValueError("email required")

    for udoc in db.collection(USERS_COLLECTION).where("email", "==", email).limit(20).stream():
        role = (udoc.to_dict() or {}).get("role")
        if role in ("counselor", "admin"):
            return udoc.id, role or "counselor"

    try:
        auth_user = fb_auth.get_user_by_email(email)
        uid = auth_user.uid
    except Exception as exc:
        raise ValueError(f"상담사 사용자를 찾을 수 없음 ({email}): {exc}") from exc

    user_doc = db.collection(USERS_COLLECTION).document(uid).get()
    if not user_doc.exists:
        raise ValueError(f"Firestore users 문서 없음 ({email}, uid={uid})")

    role = (user_doc.to_dict() or {}).get("role")
    if role not in ("counselor", "admin"):
        raise ValueError(f"상담사/admin 역할이 아님 ({email}, role={role})")

    return uid, role or "counselor"


def grant_for_email(db, email: str, amount: int, *, reason: str, actor: str) -> dict:
    uid, role = resolve_counselor_uid(db, email)
    before = get_balance(db, uid)
    result = grant_credits(
        db,
        uid,
        amount,
        reason=reason,
        actor_uid=actor,
        metadata={"source": "admin_script", "email": email.strip().lower(), "role": role},
    )
    after = result.get("balance", get_balance(db, uid))
    print(f"OK {email} uid={uid} role={role} before={before} granted={amount} after={after}")
    return result


def main() -> int:
    parser = argparse.ArgumentParser(description="Grant counselor assessment credits by email")
    parser.add_argument("emails", nargs="+", help="Counselor email(s)")
    parser.add_argument("--amount", type=int, default=100000, help="Credits to grant (default 100000)")
    parser.add_argument("--reason", default="admin_grant", help="Ledger reason")
    parser.add_argument("--actor", default="admin_script", help="Ledger actor uid")
    args = parser.parse_args()

    if args.amount <= 0 or args.amount > 100000:
        print("amount must be between 1 and 100000", file=sys.stderr)
        return 1

    db = get_firestore()
    failed = 0
    for raw in args.emails:
        email = raw.strip().lower()
        if email == "--amount" or email.startswith("--"):
            continue
        try:
            grant_for_email(db, email, args.amount, reason=args.reason, actor=args.actor)
        except Exception as exc:
            failed += 1
            print(f"FAIL {email}: {exc}", file=sys.stderr)
    return 1 if failed else 0


if __name__ == "__main__":
    raise SystemExit(main())

File: scripts/grant-counselor-credits-by-email.ts
/**
 * 상담사 이메일로 검사 크레딧 지급 (Firestore 직접)
 *
 * npx tsx scripts/grant-counselor-credits-by-email.ts jomiga@naver.com jomigata@naver.com --amount 100000
 */
import { getAuth } from 'firebase-admin/auth';
import { FieldValue, type Firestore } from 'firebase-admin/firestore';

const USERS_COLLECTION = 'users';
const COUNSELOR_CREDITS_COLLECTION = 'counselorCredits';
const COUNSELOR_CREDIT_LEDGER_COLLECTION = 'counselorCreditLedger';

async function getDb() {
  process.env.NODE_ENV = process.env.NODE_ENV || 'development';
  const mod = await import('../src/lib/firebase-admin.ts');
  if (!mod.adminDb) {
    throw new Error('Firebase Admin DB 초기화 실패 — FIREBASE_ADMIN_* 또는 GOOGLE_APPLICATION_CREDENTIALS 확인');
  }
  return mod.adminDb;
}

async function resolveCounselorUid(email: string): Promise<{ uid: string; role: string }> {
  const normalized = email.trim().toLowerCase();
  const auth = getAuth();
  const user = await auth.getUserByEmail(normalized);
  const db = await getDb();
  const snap = await db.collection(USERS_COLLECTION).doc(user.uid).get();
  if (!snap.exists) {
    throw new Error(`Firestore users 문서 없음 (${normalized})`);
  }
  const role = String(snap.data()?.role || '');
  if (role !== 'counselor' && role !== 'admin') {
    throw new Error(`상담사/admin 역할이 아님 (${normalized}, role=${role})`);
  }
  return { uid: user.uid, role };
}

async function getBalance(db: Firestore, uid: string): Promise<number> {
  const snap = await db.collection(COUNSELOR_CREDITS_COLLECTION).doc(uid).get();
  if (!snap.exists) return 0;
  const raw = snap.data()?.balance;
  const n = typeof raw === 'number' ? raw : Number(raw);
  return Number.isFinite(n) ? Math.max(0, n) : 0;
}

async function grantCredits(
  db: Firestore,
  uid: string,
  amount: number,
  email: string,
  role: string,
): Promise<number> {
  const ref = db.collection(COUNSELOR_CREDITS_COLLECTION).doc(uid);
  const before = await getBalance(db, uid);
  const after = before + amount;
  await ref.set(
    {
      counselorUid: uid,
      balance: after,
      updatedAt: FieldValue.serverTimestamp(),
    },
    { merge: true },
  );
  await db.collection(COUNSELOR_CREDIT_LEDGER_COLLECTION).add({
    counselorUid: uid,
    delta: amount,
    balanceAfter: after,
    reason: 'admin_grant',
    actorUid: 'admin_script',
    metadata: { source: 'grant_counselor_credits_by_email', email, role },
    createdAt: FieldValue.serverTimestamp(),
  });
  return after;
}

async function main() {
  const args = process.argv.slice(2);
  let amount = 100000;
  const emails: string[] = [];
  for (let i = 0; i < args.length; i += 1) {
    if (args[i] === '--amount' && args[i + 1]) {
      amount = Number(args[i + 1]);
      i += 1;
      continue;
    }
    emails.push(args[i]);
  }
  if (!emails.length) {
    console.error('Usage: npx tsx scripts/grant-counselor-credits-by-email.ts <email...> [--amount N]');
    process.exit(1);
  }
  if (!Number.isFinite(amount) || amount <= 0 || amount > 100000) {
    console.error('amount must be 1..100000');
    process.exit(1);
  }

  const db = await getDb();
  let failed = 0;
  for (const email of emails) {
    try {
      const { uid, role } = await resolveCounselorUid(email);
      const before = await getBalance(db, uid);
      const after = await grantCredits(db, uid, amount, email.trim().toLowerCase(), role);
      console.log(`OK ${email} uid=${uid} role=${role} before=${before} granted=${amount} after=${after}`);
    } catch (err) {
      failed += 1;
      console.error(`FAIL ${email}:`, err instanceof Error ? err.message : err);
    }
  }
  process.exit(failed ? 1 : 0);
}

void main();
